lib/SPIRV/SPIRVToOCL20.cpp

```diff
@@ -128,9 +142,63 @@ void SPIRVToOCL20::visitCallSPIRVControlBarrier(CallInst *CI) {
       &Attrs);
 }
 
+Instruction *SPIRVToOCL20::mutateAtomicName(CallInst *CI, Op OC) {
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+  return mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        return OCLSPIRVBuiltinMap::rmap(OC);
+      },
+      &Attrs);
+}
+
 Instruction *SPIRVToOCL20::visitCallSPIRVAtomicBuiltin(CallInst *CI, Op OC) {
+  CallInst *CIG = mutateCommonAtomicArguments(CI, OC);
+
+  Instruction *NewCI = nullptr;
+  switch (OC) {
+  case OpAtomicIIncrement:
+  case OpAtomicIDecrement:
+    NewCI = visitCallSPIRVAtomicIncDec(CIG, OC);
+    break;
+  case OpAtomicCompareExchange:
+  case OpAtomicCompareExchangeWeak:
+    NewCI = visitCallSPIRVAtomicCmpExchg(CIG, OC);
+    break;
+  default:
+    NewCI = mutateAtomicName(CIG, OC);
+  }
+
+  return NewCI;
+}
+
+Instruction *SPIRVToOCL20::visitCallSPIRVAtomicIncDec(CallInst *CI, Op OC) {
   AttributeList Attrs = CI->getCalledFunction()->getAttributes();
-  CallInst *CiFixedAs = mutateCallInstOCL(
+  return mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        // Since OpenCL 2.0 doesn't have atomic_inc and atomic_dec builtins,
+        // we translate these instructions to atomic_fetch_add_explicit and
+        // atomic_fetch_sub_explicit OpenCL 2.0 builtins with "operand" argument
+        // = 1.
+        auto Name = OCLSPIRVBuiltinMap::rmap(
+            OC == OpAtomicIIncrement ? OpAtomicIAdd : OpAtomicISub);
+        auto Ptr = findFirstPtr(Args);
+        Type *ValueTy =
+            cast<PointerType>(Args[Ptr]->getType())->getElementType();
+        assert(ValueTy->isIntegerTy());
+        Args.push_back(llvm::ConstantInt::get(ValueTy, 1));
+        std::swap(Args[Ptr + 1], Args.back());
```

Comment:
I guess the test is wrong because of this swap

Reply:
Yes, that was it. Thanks :)